genet/schedule_elements.py

```diff
@@ -972,6 +1012,125 @@ def _build_graph(self, services):
         schedule_graph.graph['services'] = graph_services
         return schedule_graph
 
+    def generate_vehicles(self):
+        """
+        Generate vehicles for the Schedule. Returns dictionary of vehicle IDs from Route objects, mapping them to
+        vehicle types in vehicle_types. Looks like this:
+            {veh_id : {'type': 'bus'}}
+        Generates itself from the vehicles IDs which exist in Routes, maps to the mode of the Route.
+        :return:
+        """
+        if self:
+            # generate vehicles using Services and Routes upon init
+            df = self.route_trips_to_dataframe()[['route_id', 'vehicle_id']]
+            df['type'] = df.apply(
+                lambda x: self._graph.graph['routes'][x['route_id']]['mode'], axis=1)
+            df = df.drop(columns='route_id')
+            # check mode consistency
+            vehicles_to_modes = df.groupby('vehicle_id').apply(lambda x: list(x['type'].unique()))
+            if (vehicles_to_modes.str.len() > 1).any():
+                # there are vehicles which are shared across routes with different modes
```

Comment:
Does this ever happen? Would hope not but good to have a catch for it.
If it does with our ITO data it is something we should feed back to them.

Reply:
so when reading GTFS I give unique vehicles to each trip so this doesn't happen, is there data in the GTFS that could serve as a vehicle? I added this because 
1. people might like to generate vehicles that are shared between trips to make things more realistic, and it's really easily done with the `route_trips_to_dataframe` and `set_route_trips_dataframe` methods
2. People could get their vehicle ids muddled when adding routes and services

Reply:
Got it